--- a/drivers/mtd/mtdpart.c
+++ b/drivers/mtd/mtdpart.c
@@ -614,6 +614,7 @@ int mtd_add_partition(struct mtd_info *m
 	ret = add_mtd_device(&new->mtd);
 	if (ret)
 		goto err_remove_part;
+        mtd_partition_split(master, new);
 
 	mtd_add_partition_attrs(new);
 
@@ -693,8 +694,9 @@ int add_mtd_partitions(struct mtd_info *
 
 			free_partition(slave);
 			goto err_del_partitions;
		}
 
+                mtd_partition_split(master, slave);
 		mtd_add_partition_attrs(slave);
 
 		cur_offset = slave->offset + slave->mtd.size;
