--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3438,9 +3438,27 @@ begin:
 	tx.skb = skb;
 	tx.sdata = vif_to_sdata(info->control.vif);
 
-	if (txq->sta)
+	if (txq->sta) {
 		tx.sta = container_of(txq->sta, struct sta_info, sta);
 
+        /*
+ 	 * Drop unicast frames to unauthorised stations unless they are
+ 	 * EAPOL frames from the local station.
+ 	 */
+ 	if (unlikely(!ieee80211_vif_is_mesh(&tx.sdata->vif) &&
+ 		     tx.sdata->vif.type != NL80211_IFTYPE_OCB &&
+ 		     !is_multicast_ether_addr(hdr->addr1) &&
+ 		     !test_sta_flag(tx.sta, WLAN_STA_AUTHORIZED) &&
+ 		     (!(info->control.flags &
+ 			IEEE80211_TX_CTRL_PORT_CTRL_PROTO) ||
+ 		      !ether_addr_equal(tx.sdata->vif.addr,
+ 					hdr->addr2)))) {
+ 		I802_DEBUG_INC(local->tx_handlers_drop_unauth_port);
+ 		ieee80211_free_txskb(&local->hw, skb);
+ 		goto begin;
+ 		}
+ 	}
+
 	/*
 	 * The key can be removed while the packet was queued, so need to call
 	 * this here to get the current key.
